### PR TITLE
Rework Behat testing from classes/ids to attributes for login and registration pages

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/UiBundle/Form/Type/SecurityLoginType.php [hunk 1]
```diff
@@ -21,28 +21,25 @@
...
 {
-    /**
-     * {@inheritdoc}
-     */
     public function buildForm(FormBuilderInterface $builder, array $options): void
     {
         $builder
             ->add('_username', TextType::class, [
                 'label' => 'sylius.form.login.username',
+                'attr' => ['data-login-username' => ''],
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Form/Type/SecurityLoginType.php]: I think we should add it to the template instead.

`form_row(form._username, {'attr': {'data-login-username': ''}})`

In order to make it more extendable, let's add a new Twig function `sylius_form_test_attribute(attribute)` which would return `['attr' => [$attribute => '']]` in test environment and `[]` in any other environment.

This way you can define the additional attributes in an easy way:

`form_row(form._username, sylius_form_test_attribute('data-login-username'))`

Or in case that there's some configuration already defined:

`form_row(form._username, sylius_form_test_attribute('data-login-username')|merge({'other': 'config'}))`

Then the attributes could be removed from the PHP form types.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/resetPassword.html.twig [hunk 2]
```diff
@@ -18,7 +18,9 @@
                 {{ form_row(form.password.first, {'label': 'sylius.form.user_reset_password.new'|trans}) }}
                 {{ form_row(form.password.second, {'label': 'sylius.form.user_reset_password.confirmation'|trans}) }}
                 {{ form_row(form._token) }}
-                <button type="submit" class="ui fluid large primary submit button">{{ 'sylius.ui.reset'|trans }}</button>
+                <button type="submit" class="ui fluid large primary submit button" data-login-password-reset-button>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/resetPassword.html.twig]:
```suggestion
                <button type="submit" class="ui fluid large primary submit button" {{ sylius_test_attribute(data-login-password-reset-button) }}>
```

Comment (AdamKasp) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Account/resetPassword.html.twig]: here and everywhere

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/requestPasswordReset.html.twig [hunk 3]
```diff
@@ -22,7 +22,9 @@
                 {{ sonata_block_render_event('sylius.shop.request_password_reset_token.form', {'form': form}) }}
 
                 {{ form_row(form._token) }}
-                <button type="submit" class="ui fluid large primary submit button">{{ 'sylius.ui.reset'|trans }}</button>
+                <button type="submit" class="ui fluid large primary submit button" {{ sylius_test_attribute('data-request-password-reset-button') }}>
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/requestPasswordReset.html.twig]: I would be for adding `data-` prefix in extension, so you could have `{{ sylius_test_attribute('request-password-reset-button') }}`

File: src/Sylius/Bundle/UiBundle/Form/Type/SecurityLoginType.php (hunk 1 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/UiBundle/Form/Type/SecurityLoginType.php]: You a right it is great idea 👍

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/requestPasswordReset.html.twig [hunk 4]
```diff
@@ -17,12 +17,14 @@
             </h1>
             <div class="ui segment">
                 {{ form_start(form, {'action': path('sylius_shop_request_password_reset_token'), 'attr': {'class': 'ui large loadable form', 'novalidate': 'novalidate'}}) }}
-                {{ form_row(form.email) }}
+                {{ form_row(form.email, sylius_data_test_attribute_array('reset-email')) }}
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/requestPasswordReset.html.twig]:
```suggestion
                {{ form_row(form.email, sylius_test_form_attribute('reset-email')) }}
```
It's an array, but it's highly specific for forms and should only be used there - hence my proposition to change the name.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/requestPasswordReset.html.twig [hunk 5]
```diff
@@ -17,12 +17,14 @@
...
 
                 {{ sonata_block_render_event('sylius.shop.request_password_reset_token.form', {'form': form}) }}
 
                 {{ form_row(form._token) }}
-                <button type="submit" class="ui fluid large primary submit button">{{ 'sylius.ui.reset'|trans }}</button>
+                <button type="submit" class="ui fluid large primary submit button" {{ sylius_data_test_attribute_name('request-password-reset-button') }}>
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/requestPasswordReset.html.twig]:
```suggestion
                <button type="submit" class="ui fluid large primary submit button" {{ sylius_test_html_attribute('request-password-reset-button') }}>
```

Going further with the same logic as in the comment above, this will mark it as specific for HTML and also make the function name shorter.

File: src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeArrayExtension.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * @return TwigFunction[]
+     */
+    public function getFunctions(): array
+    {
+        return [
+            new TwigFunction(
+                'sylius_data_test_attribute_array',
+                function (?string $data = null, ?string $attributeValue = null): array {
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeArrayExtension.php]:
```suggestion
                function (string $name, ?string $value = null): array {
```

`$data` should be always passed.

I'm also wondering about `$name` for that instead of `$data`.

File: src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeArrayExtension.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                    if (strpos($this->env, 'test') !== false) {
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeArrayExtension.php]:
```suggestion
                    if (strpos($this->env, 'test') === 0) {
```

I would go for checking if `test` starts the environment string instead of checking whether `test` could be found anywhere in that string.

File: src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeArrayExtension.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                        return ['attr' => ['data-test-'.$data => sprintf('"%s"', $attributeValue)]];
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeArrayExtension.php]:
```suggestion
                        return ['attr' => ['data-test-'.$data => (string) $value];
```

Comment (pamil) [reply src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeArrayExtension.php]: I think we don't need `"` here.

File: src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * @return TwigFunction[]
+     */
+    public function getFunctions(): array
+    {
+        return [
+            new TwigFunction(
+                'sylius_data_test_attribute_name',
+                function (?string $data = null, ?string $attributeValue = null): ?string {
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php]:
```suggestion
                function (string $name, ?string $value = null): string {
```

File: src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                    if (strpos($this->env, 'test') !== false) {
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php]: Same as above.

File: src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                        return sprintf('data-test-%s="%s"', $data, $attributeValue);
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php]: Let's cast `$attributeValue` to string explicitly - `(string) $attributeValue`.

File: src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+                    }
+
+                    return null;
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/DataTestAttributeNameExtension.php]: It should rather return an empty string here.

File: src/Sylius/Behat/Element/Shop/Account/RegisterElement.php [hunk 13]
```diff
@@ -51,43 +51,45 @@ public function getEmail(): string
...
     protected function getDefinedElements(): array
     {
         return array_merge(parent::getDefinedElements(), [
-            'email' => '#sylius_customer_registration_email',
-            'first_name' => '#sylius_customer_registration_firstName',
-            'last_name' => '#sylius_customer_registration_lastName',
-            'password' => '#sylius_customer_registration_user_plainPassword_first',
-            'password_verification' => '#sylius_customer_registration_user_plainPassword_second',
-            'phone_number' => '#sylius_customer_registration_phoneNumber',
+            'email' => '[data-test-email]',
```
Comment (lchrusciel) [inline src/Sylius/Behat/Element/Shop/Account/RegisterElement.php]: Maybe we can process some helper here? It could be done in the next iteration.

Comment (lchrusciel) [reply src/Sylius/Behat/Element/Shop/Account/RegisterElement.php]: Something like this:
```suggestion
            'email' => TestAttribute::string('email'),
```

File: src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+     * @return TwigFunction[]
+     */
+    public function getFunctions(): array
+    {
+        return [
+            new TwigFunction(
+                'sylius_test_form_attribute',
+                function (string $name, ?string $value = null): array {
+                    if (strpos($this->env, 'test') === 0) {
+                        return ['attr' => ['data-test-'.$name => (string) $value]];
```
Comment (lchrusciel) [inline src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]:
```suggestion
                        return ['attr' => ['data-test-' . $name => (string) $value]];
```

Comment (lchrusciel) [reply src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]: But in the case, we would like to introduce helper class, we can do
```suggestion
                        return ['attr' => TestAttribure::array($name, $value)];
```

File: src/Sylius/Bundle/UiBundle/Twig/TestHtmlAttributeExtension.php [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+     * @return TwigFunction[]
+     */
+    public function getFunctions(): array
+    {
+        return [
+            new TwigFunction(
+                'sylius_test_html_attribute',
+                function (string $name, ?string $value = null): string {
+                    if (strpos($this->env, 'test') === 0) {
+                        return sprintf('data-test-%s="%s"', $name, (string) $value);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/UiBundle/Twig/TestHtmlAttributeExtension.php]:
```suggestion
                        return TestAttribute::string($name, $value);
```

File: src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php (hunk 14 above)
Comment (pamil) [reply src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]: I'd say a helper class used only in one place without any possibility to replace is not a good idea, especially for maintainers.

File: src/Sylius/Behat/Element/Shop/Account/RegisterElement.php (hunk 13 above)
Comment (pamil) [reply src/Sylius/Behat/Element/Shop/Account/RegisterElement.php]: I get the idea here but I'd leave it for later when we will have more experience with the new system so that we create right abstraction layer.

File: src/Sylius/Behat/Context/Ui/Shop/LoginContext.php [hunk 16]
```diff
@@ -117,7 +117,7 @@ public function iSpecifyTheEmail($email = null)
      * @When I specify the password as :password
      * @When I do not specify the password
      */
-    public function iSpecifyThePasswordAs($password = null)
+    public function iSpecifyThePasswordAs(string $password = null): void
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Shop/LoginContext.php]:
```suggestion
    public function iSpecifyThePasswordAs(?string $password = null): void
```

File: src/Sylius/Behat/Context/Ui/Shop/LoginContext.php [hunk 17]
```diff
@@ -126,7 +126,7 @@ public function iSpecifyThePasswordAs($password = null)
      * @When I specify my new password as :password
      * @When I do not specify my new password
      */
-    public function iSpecifyMyNewPassword(string $password = null)
+    public function iSpecifyMyNewPassword(string $password = null): void
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Shop/LoginContext.php]:
```suggestion
    public function iSpecifyMyNewPassword(?string $password = null): void
```

File: src/Sylius/Behat/Context/Ui/Shop/LoginContext.php [hunk 18]
```diff
@@ -135,7 +135,7 @@ public function iSpecifyMyNewPassword(string $password = null)
      * @When I confirm my new password as :password
      * @When I do not confirm my new password
      */
-    public function iConfirmMyNewPassword(string $password = null)
+    public function iConfirmMyNewPassword(string $password = null): void
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Shop/LoginContext.php]:
```suggestion
    public function iConfirmMyNewPassword(?string $password = null): void
```

File: src/Sylius/Behat/Context/Ui/Shop/LoginContext.php [hunk 19]
```diff
@@ -75,31 +75,31 @@ public function __construct(
...
+    public function iFollowLinkOnMyEmailToResetPassword(UserInterface $user): void
     {
         $this->resetPasswordPage->open(['token' => $user->getPasswordResetToken()]);
     }
 
     /**
      * @When I specify the username as :username
      */
-    public function iSpecifyTheUsername($username = null)
+    public function iSpecifyTheUsername(string $username = null): void
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Shop/LoginContext.php]:
```suggestion
    public function iSpecifyTheUsername(?string $username = null): void
```

File: src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php (hunk 14 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]: It won't be used only in one place. This class should be used in all behat scenarios. What I would like to achieve, is to have this string (and array) building logic in one class. 

Injecting it would be even better of course. The question is if it still is worth it when we will have to inject this service to every page. On the other hand, we can just extend parent class and add it as a protected field.

File: src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFunction;
+
+final class TestFormAttributeExtension extends AbstractExtension
+{
+    /** @var string */
+    private $env;
+
+    public function __construct(string $env)
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]: Btw. I would go for `$environment`, no need for abbreviations in the code.

File: src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php (hunk 14 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]: Just to be clear. The same service (or helper class) should be used here and here: https://github.com/Sylius/Sylius/pull/10883/files#r352113248

Comment (pamil) [reply src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]: The output is different here:
 - for rendering we need:
   - HTML attribute: `data-test-key` or `data-test-key="value"`
   - form attribute: `['attr' => ['data-test-key' => '']]` or `['attr' => ['data-test-key' => 'value']]`
 - for matching we need:
   - either purely key-based: `[data-test-key]`
   - or basing on key and value: `[data-test-key="value"]`

If we go with static class, I'd only implement two methods:
 - one that creates the identifier, passing `key` as a parameter would return string `data-test-key`
 - one that creates the matcher, passing `key` as a parameter would return string  `[data-test-key]` and passing `key` and `value` would return string `[data-test-key="value"]`

I don't see where method returning array `['data-test-'.$key => (string) $value]` for `$key, $value` could be reused.

Comment (lchrusciel) [reply src/Sylius/Bundle/UiBundle/Twig/TestFormAttributeExtension.php]: I thought about putting all creation methods into one class, but anything will be better than writing strings

**Comment by lchrusciel** [comment]
Thank you, Tomasz! :1st_place_medal:

File: src/Sylius/Bundle/ShopBundle/Resources/views/Menu/_security.html.twig [hunk 21]
```diff
@@ -2,7 +2,7 @@
     {% if is_granted('ROLE_USER') %}
         <div class="item">{{ 'sylius.ui.hello'|trans }} {{ app.user.customer.fullName }}!</div>
         <a href="{{ path('sylius_shop_account_dashboard') }}" class="item">{{ 'sylius.ui.my_account'|trans }}</a>
-        <a href="{{ path('sylius_shop_logout') }}" class="item sylius-logout-button">{{ 'sylius.ui.logout'|trans }}</a>
+        <a href="{{ path('sylius_shop_logout') }}" class="item" {{ sylius_test_html_attribute('logout-button') }}>{{ 'sylius.ui.logout'|trans }}</a>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Menu/_security.html.twig]:
```suggestion
        <a href="{{ path('sylius_shop_logout') }}" class="item  sylius-logout-button" {{ sylius_test_html_attribute('logout-button') }}>{{ 'sylius.ui.logout'|trans }}</a>
```
Please revert this change
